modules/matterhorn-capture-admin-service-impl/src/main/java/org/opencastproject/capture/admin/impl/CaptureAgentStateServiceImpl.java

```diff
@@ -505,35 +502,35 @@ private void updateAgentInDatabase(AgentImpl agent, boolean updateFromCache) {
     EntityManager em = null;
     EntityTransaction tx = null;
     try {
-      em = emf.createEntityManager();
-      tx = em.getTransaction();
-      tx.begin();
-      AgentImpl existing = getAgentEntity(agent.getName(), agent.getOrganization(), em);
-
+      //This is the cached last-heard-from time
+      Long cached = -1L;
       // Update the last seen property from the agent cache
-      if (existing != null && updateFromCache) {
+      if (updateFromCache) {
         try {
-          Tuple3<String, Properties, Long> cachedAgent = getAgentFromCache(existing.getName(),
-                  existing.getOrganization());
-          if (agent != null && cachedAgent != null) {
-            agent.setLastHeardFrom(cachedAgent.getC());
-          }
+          cached = getAgentFromCache(agent.getName(), agent.getOrganization()).getC();
         } catch (NotFoundException e) {
           // That's fine
         }
       }
 
-      if (existing == null) {
-        em.persist(agent);
-      } else {
+      //Make sure the last heard from value is as large as possible
+      Long larger = Math.max(cached, agent.getLastHeardFrom());
```

Comment:
_larger_ is a quite unspecific variable name. Also, the comment "is as large as possible" is not correct.

Maybe just omit those two lines and set existing.setLastHeardFrom(Math.max(lastHeardFromCached, agent.getLastHeardFrom()); ?